nupic/research/frameworks/vernon/mixins/extra_validations.py

```diff
@@ -0,0 +1,140 @@
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy as np
+
+from nupic.research.frameworks.pytorch.model_utils import aggregate_eval_results
+
+__all__ = [
+    "ExtraValidations"
+]
+
+
+class ExtraValidations:
+    """
+    Perform validations mid-epoch. Requires StepBasedLogging core mixin.
+    """
+    def setup_experiment(self, config):
+        """
+        :param config: Dictionary containing the configuration parameters
+
+            - extra_validations_per_epoch: number of additional validations to
+                                           perform mid-epoch. Additional
+                                           validations are distributed evenly
+                                           across training batches.
+        """
+        super().setup_experiment(config)
+
+        assert hasattr(self, "current_timestep"), (
+            "Must use StepBasedLogging or similar extension"
+        )
+
+        # A list of [(timestep, result), ...] for the current epoch.
+        self.extra_val_results = []
+
+        extra_validations = config.get("extra_validations_per_epoch", 0)
+        batches_to_validate = np.linspace(
+            min(self.total_batches, self.batches_in_epoch),
+            0,
+            1 + extra_validations,
+            endpoint=False
+        )[::-1].round().astype("int").tolist()
+        self.additional_batches_to_validate = batches_to_validate[:-1]
+        if extra_validations > 0:
+            self.logger.info(
+                f"Extra validations per epoch: {extra_validations}, "
+                f"batch indices: {self.additional_batches_to_validate}")
+
+    def run_epoch(self):
+        ret = super().run_epoch()
+        ret["extra_val_results"] = self.extra_val_results
+        self.extra_val_results = []
+        return ret
+
+    def post_batch(self, batch_idx, **kwargs):
+        super().post_batch(batch_idx=batch_idx, **kwargs)
+        validate = (batch_idx in self.additional_batches_to_validate
+                    and self.current_epoch in self.epochs_to_validate)
+        if validate:
+            result = self.validate()
+            self.extra_val_results.append(
+                (self.current_timestep, result)
+            )
+            self.model.train()
+
+    @classmethod
+    def expand_result_to_time_series(cls, result, config):
```

Comment:
Could you add a docstring to this? Specifically, it could be helpful to help distinguish it from `StepBasedLogging. expand_result_to_time_series`.

Reply:
I was a bit confused by the term `time series`, at least for me it relates for a method to handle datasets with a temporal dimension (like in RNNs / HTMs).

Reply:
I added some code comments. (At the moment, I'm not quite sure how I should think about docstrings on an overriden method. Docstrings on interfaces are surfaced to IDEs, etc., so I'm making sure the interfaces have good docstrings.)

I'll think about new terminology in the future... the "time series" terminology has been in this code for many months. (And this change is already big.)